### PR TITLE
[Core] Add shape derivatives to ElementSizeCalculator

### PR DESCRIPTION
**Pull request by sunethwarna**
**Description**
This PR adds shape derivative computations for minimum and average element size methods.

**Changelog**
- Added Minimum and Average element size derivative computation methods
- Added finite difference tests

### CODE REVIEW

**Comment by philbucher** [comment]
looks ok to me, but I am not really familiar with this part of the code

@rubenzorrilla can you take a look?

**Review by rubenzorrilla — COMMENTED** [review]
I'm not familar with this element size calculation method. Can you please document it in the code?

File: kratos/utilities/element_size_calculator.h [hunk 1]
```diff
@@ -116,6 +138,67 @@ class KRATOS_API(KRATOS_CORE) ElementSizeCalculator {
...
+        const double VyDerivative,
+        const double Ny,
+        const double NyDerivative)
+    {
+        return 2 * (Vx * Nx + Vy * Ny) *
+                   (VxDerivative * Nx + Vx * NxDerivative + VyDerivative * Ny +
+                    Vy * NyDerivative) / (Nx * Nx + Ny * Ny) -
+               std::pow((Vx * Nx + Vy * Ny) / (Nx * Nx + Ny * Ny), 2) *
+                   (2 * Nx * NxDerivative + 2 * Ny * NyDerivative);
+    }
```
Comment (rubenzorrilla) [inline kratos/utilities/element_size_calculator.h:160]: Why is this implemented in the header?

File: kratos/utilities/element_size_calculator.h [hunk 2]
```diff
@@ -116,6 +138,67 @@ class KRATOS_API(KRATOS_CORE) ElementSizeCalculator {
...
+        const double VzDerivative,
+        const double Nz,
+        const double NzDerivative)
+    {
+        return 2 * (Vx * Nx + Vy * Ny + Vz * Nz) *
+                   (VxDerivative * Nx + Vx * NxDerivative + VyDerivative * Ny +
+                    Vy * NyDerivative + VzDerivative * Nz + Vz * NzDerivative) / (Nx * Nx + Ny * Ny + Nz * Nz) -
+               std::pow((Vx * Nx + Vy * Ny + Vz * Nz) / (Nx * Nx + Ny * Ny + Nz * Nz), 2) *
+                   (2 * Nx * NxDerivative + 2 * Ny * NyDerivative + 2 * Nz * NzDerivative);
+    }
```
Comment (rubenzorrilla) [inline kratos/utilities/element_size_calculator.h:181]: Same.

File: kratos/utilities/element_size_calculator.h [hunk 3]
```diff
@@ -116,6 +138,67 @@ class KRATOS_API(KRATOS_CORE) ElementSizeCalculator {
...
+
+    static inline double NormSquareDerivative(
+        const array_1d<double, 3>& rValue,
+        const array_1d<double, 3>& rValueDerivative)
+    {
+        return 2 * inner_prod(rValue, rValueDerivative);
+    }
```
Comment (rubenzorrilla) [inline kratos/utilities/element_size_calculator.h:188]: Same.

File: kratos/utilities/element_size_calculator.h [hunk 4]
```diff
@@ -116,6 +138,67 @@ class KRATOS_API(KRATOS_CORE) ElementSizeCalculator {
...
+
+    static inline double EdgeLengthDerivative(
+        const unsigned int DerivativeNodeIndex,
+        const unsigned int DerivativeDirectionIndex,
+        const unsigned int NodeIndexA,
+        const unsigned int NodeIndexB,
+        const unsigned int EdgeDirection)
+    {
+        return ((DerivativeNodeIndex == NodeIndexA) - (DerivativeNodeIndex == NodeIndexB)) * (DerivativeDirectionIndex == EdgeDirection);
+    }
```
Comment (rubenzorrilla) [inline kratos/utilities/element_size_calculator.h:198]: Same.

**Comment by rubenzorrilla** [comment]
OK. I think now I understand. You are adding the minimum and average element sizes formulas derivatives. As this is a way specific operation, I think it would be valuable to have some theory hints in the documentation. I guess this is the space derivative of the minimum and average element size calculation formulas. Am I wrong?

File: kratos/utilities/element_size_calculator.h (hunk 1 above)
Comment (sunethwarna) [reply kratos/utilities/element_size_calculator.h:160]: I wanted to have those methods inlined. Thats the only reason. Is there any other way around it?

**Comment by sunethwarna** [comment]
You are correct. I am not sure how to add the formulations. How do you propose to add the formulations?

**Review by rubenzorrilla — APPROVED** [review]

**Pull request merged**